### PR TITLE
ISPN-8963 Remote getCacheNames

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-8963 RemoteCacheManager.getCacheNames()
https://issues.jboss.org/browse/ISPN-8328 SpringRemoteCacheManager.getCacheNames()

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java [hunk 1]
```diff
@@ -153,13 +157,26 @@ public RemoteCacheManager() {
...
    public <K, V> RemoteCache<K, V> getCache(String cacheName) {
       return getCache(cacheName, configuration.forceReturnValues());
    }
 
+   @Override
+   public Set<String> getCacheNames() {
+      OperationsFactory operationsFactory = new OperationsFactory(channelFactory, codec, configuration);
+      String names = await(operationsFactory.newAdminOperation("@@cache@names", Collections.emptyMap()).execute());
+      Set<String> cacheNames = new HashSet<>();
+      Pattern pattern = Pattern.compile("(?:\")([^\"]*)(?:\",?)");
```
Comment (alanfx) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java]: Can you add a comment about what this RegEx is matching?

File: commons/src/main/java/org/infinispan/commons/api/BasicCacheContainer.java [hunk 2]
```diff
@@ -49,4 +51,19 @@
...
+
+   /**
+    * This method returns a collection of caches names which contains the
+    * caches that have been defined via XML or programmatically, and the
+    * caches that have been created at runtime via this cache manager
+    * instance.
+    *
+    * If no named caches are registered or no caches have been created, this
+    * method returns an empty set.  The default cache is never included in this
+    * set of cache names, as well as caches for internal-only use {@link org.infinispan.registry.InternalCacheRegistry}
```
Comment (danberindei) [inline commons/src/main/java/org/infinispan/commons/api/BasicCacheContainer.java]: No longer true with `default-cache="..."`?

**Comment by danberindei** [comment]
Looks good to me, apart from the outdated javadoc.

File: server/core/src/main/java/org/infinispan/server/core/admin/embeddedserver/CacheNamesTask.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+   public Set<String> getParameters() {
+      return Collections.emptySet();
+   }
+
+   @Override
+   protected byte[] execute(EmbeddedCacheManager cacheManager, Map<String, String> parameters, EnumSet<CacheContainerAdmin.AdminFlag> adminFlags) {
+      Set<String> cacheNames = cacheManager.getCacheNames();
+      StringBuilder sb = new StringBuilder("[");
+      for(String s : cacheNames) {
+         if (sb.length() > 0)
```
Comment (anistor) [inline server/core/src/main/java/org/infinispan/server/core/admin/embeddedserver/CacheNamesTask.java]: ```sb.length() > 1```, please

**Comment by tristantarrant** [comment]
Addressed all comments

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java [hunk 4]
```diff
@@ -153,13 +157,27 @@ public RemoteCacheManager() {
...
       return getCache(cacheName, configuration.forceReturnValues());
    }
 
+   @Override
+   public Set<String> getCacheNames() {
+      OperationsFactory operationsFactory = new OperationsFactory(channelFactory, codec, configuration);
+      String names = await(operationsFactory.newAdminOperation("@@cache@names", Collections.emptyMap()).execute());
+      Set<String> cacheNames = new HashSet<>();
+      // Simple pattern that matches the result which is represented as a JSON string array, e.g. ["cache1","cache2"]
+      Pattern pattern = Pattern.compile("(?:\")([^\"]*)(?:\",?)");
```
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java]: This is could be extracted as a constant

File: spring/spring4/spring4-remote/src/test/java/org/infinispan/spring/provider/SpringRemoteCacheManagerTest.java [hunk 5]
```diff
@@ -50,7 +53,9 @@ protected EmbeddedCacheManager createCacheManager() throws Exception {
 
    @BeforeClass
    public void setupRemoteCacheFactory() {
-      hotrodServer = HotRodTestingUtil.startHotRodServer(cacheManager, 19722);
+      HotRodServerConfigurationBuilder serverBuilder = new HotRodServerConfigurationBuilder();
+      serverBuilder.adminOperationsHandler(new EmbeddedServerAdminOperationHandler());
+      hotrodServer = HotRodTestingUtil.startHotRodServer(cacheManager, 19722, serverBuilder);
```
Comment (gustavonalle) [inline spring/spring4/spring4-remote/src/test/java/org/infinispan/spring/provider/SpringRemoteCacheManagerTest.java]: A good opportunity to not hardcode the port here, but rather use ```findFreePort()```

**Comment by tristantarrant** [comment]
Addressed  @gustavonalle 's comments too

**Pull request merged**